source/OrderedTypes/InductiveTypesfromSmallGeneratedLattice.lagda

```diff
@@ -0,0 +1,2209 @@
+                  → ((x : B) → (u : x ∈ U) → f x u ∈ P x)
+                  → (b : B) → (g : (b ≤ᴮ (⋁ (𝕋 U , q ∘ 𝕋-to-carrier U))))
+                  → c-closed U f b g ∈ P b)
+                 → ((a : ⟨ L ⟩)
+                  → (b : B)
+                  → (p : (b , a) ∈ ϕ)
+                  → (f : (x : B) → (x ≤ᴮ a → x ∈ 𝓘nd))
+                  → ((x : B) → (o : x ≤ᴮ a) → f x o ∈ P x)
+                  → ϕ-closed a b p f ∈ P b)
+                 → (b : B) → (i : b ∈ 𝓘nd) → i ∈ P b
+   𝓘nd-induction = Ind-induction
+
+   𝓘nd-recursion : (P : 𝓟 {𝓣} B)
+                 → ((U : 𝓟 {𝓥} B)
+                  → (U ⊆ 𝓘nd)
+                  → (U ⊆ P)
+                  → (b : B) → (b ≤ᴮ (⋁ (𝕋 U , q ∘ 𝕋-to-carrier U)))
+                  → b ∈ P)
+                 → ((a : ⟨ L ⟩)
+                  → (b : B)
+                  → (b , a) ∈ ϕ
+                  → ((x : B) → (x ≤ᴮ a → x ∈ 𝓘nd))
+                  → ((x : B) → (x ≤ᴮ a → x ∈ P))
+                  → b ∈ P)
+                 → 𝓘nd ⊆ P
+   𝓘nd-recursion P = 𝓘nd-induction λ b → (λ _ → P b)
+
+   𝓘nd-is-initial : (P : 𝓟 {𝓣} B)
+                  → ((U : 𝓟 {𝓥} B)
+                   → (U ⊆ P)
+                   → ((b : B) → (b ≤ᴮ (⋁ (𝕋 U , q ∘ 𝕋-to-carrier U)))
+                   → b ∈ P))
+                  → ((a : ⟨ L ⟩)
+                   → (b : B)
+                   → (b , a) ∈ ϕ
+                   → ((b' : B) → (b' ≤ᴮ a → b' ∈ P)) → b ∈ P)
+                  → 𝓘nd ⊆ P
+   𝓘nd-is-initial {𝓣} P IH₁ IH₂ b b-in-𝓘nd = 𝓘nd-recursion P R S b b-in-𝓘nd
+    where
+     R : (U : 𝓟 {𝓥} B)
+       → U ⊆ 𝓘nd
+       → U ⊆ P
+       → (x : B) → x ≤ᴮ (⋁ (𝕋 U , q ∘ 𝕋-to-carrier U))
+       →  x ∈ P
+     R U C₁ C₂ x o = IH₁ U C₂ x o
+     S : (a : ⟨ L ⟩)
+       → (x : B)
+       → (x , a) ∈ ϕ
+       → ((z : B) → z ≤ᴮ a → z ∈ 𝓘nd)
+       → ((z : B) → z ≤ᴮ a → z ∈ P)
+       → x ∈ P
+     S a x p f g = IH₂ a x p g
+
+\end{code}
+
+We now consider a restricted class of inductive definitions which we will call
+local. A local inductive definition ϕ is one such that the type
+
+  Σ b ꞉ B , (Ǝ a' ꞉ ⟨ L ⟩ , (b , a') ∈ ϕ × (a' ≤ a)
+
+is small.
+
+We then define an operator parameterized by local inductive definitions
+and prove that it is monotone. Finally, we show that any monotone endo map on
+the sup-lattice corresponds to a monotone operator and local inductive
+definition.
+
+\begin{code}
+
+module local-inductive-definitions {𝓤 𝓦 𝓥 : Universe}
+                                   {B : 𝓥  ̇}
+                                   (L : Sup-Lattice 𝓤 𝓦 𝓥)
+                                   (q : B → ⟨ L ⟩)
+                                    where
+
+ open small-basis L q
+ open Joins _≤_
+ open inductive-definitions L q 
+
+ module local-from-small-basis-facts (h : is-small-basis) where
+
+  open PropositionalTruncation pt
+  open small-basis-facts h
+
+  S : (ϕ : 𝓟 {𝓤 ⊔ 𝓥} (B × ⟨ L ⟩)) → (a : ⟨ L ⟩) → 𝓤 ⊔ 𝓦 ⊔ 𝓥  ̇
+  S ϕ a = Σ b ꞉ B , (Ǝ a' ꞉ ⟨ L ⟩ , (b , a') ∈ ϕ × (a' ≤ a) holds) holds
```

Comment:
A more descriptive and less overloaded name would be desirable.

Reply:
I am struggling with naming this... do you have any suggestions?